src/compiler/transformers/esnext.ts

```diff
@@ -42,8 +42,13 @@ namespace ts {
         function flattenChain(chain: OptionalChain) {
             const links: OptionalChain[] = [chain];
             while (!chain.questionDotToken && !isTaggedTemplateExpression(chain)) {
-                chain = cast(chain.expression, isOptionalChain);
-                links.unshift(chain);
+                if (isOptionalChain(chain.expression)) {
+                    chain = chain.expression;
+                    links.unshift(chain);
+                }
+                else {
+                    return { expression: chain.expression, chain: links };
+                }

```

Comment:
Why was this change necessary? Appears unrelated to private identifier support

Reply:
It's the dread support for `--noEmitOnError false` again.

Without the change above, the cast fails on things like `this?.foo.#bar`.

After the change, we short-circuit when the cast fails, and make some silly emit that will fail at runtime. I think you can see the emit in PrivateIdentifierChain.1.js